sae_training/train_sae_on_language_model.py

```diff
@@ -95,240 +258,195 @@ def train_sae_on_language_model(
-    # save sae group to checkpoints folder
-    path = f"{sae_group.cfg.checkpoint_path}/final_{sae_group.get_name()}.pt"
+class SaveCheckpointOutput(NamedTuple):
+    path: str
+    log_feature_sparsity_path: str
+    log_feature_sparsities: list[torch.Tensor]
+
+
+def _save_checkpoint(
+    sae_group: SAEGroup,
+    train_contexts: list[SAETrainContext],
+    checkpoint_name: int | str,
+    wandb_aliases: list[str] | None = None,
+) -> SaveCheckpointOutput:
+    path = (
+        f"{sae_group.cfg.checkpoint_path}/{checkpoint_name}_{sae_group.get_name()}.pt"
+    )
     for sae in sae_group:
         sae.set_decoder_norm_to_unit_norm()
     sae_group.save_model(path)
-
+    log_feature_sparsity_path = f"{sae_group.cfg.checkpoint_path}/{checkpoint_name}_{sae_group.get_name()}_log_feature_sparsity.pt"
+    log_feature_sparsities = [
+        _log_feature_sparsity(ctx.feature_sparsity) for ctx in train_contexts
+    ]
+    torch.save(log_feature_sparsities, log_feature_sparsity_path)
     if sae_group.cfg.log_to_wandb:
         model_artifact = wandb.Artifact(
             f"{sae_group.get_name()}",
             type="model",
             metadata=dict(sae_group.cfg.__dict__),
         )
         model_artifact.add_file(path)
-        wandb.log_artifact(model_artifact, aliases=["final_model"])
-
-    # need to fix this
-    log_feature_sparsity_path = f"{sae_group.cfg.checkpoint_path}/final_{sae_group.get_name()}_log_feature_sparsity.pt"
-    log_feature_sparsity = []
-    for sae_id in range(len(sae_group)):
-        feature_sparsity = act_freq_scores[sae_id] / n_frac_active_tokens[sae_id]
-        log_feature_sparsity.append(
-            torch.log10(feature_sparsity + 1e-10).detach().cpu()
-        )
-    torch.save(log_feature_sparsity, log_feature_sparsity_path)
+        wandb.log_artifact(model_artifact, aliases=wandb_aliases)
 
-    if sae_group.cfg.log_to_wandb:
         sparsity_artifact = wandb.Artifact(
             f"{sae_group.get_name()}_log_feature_sparsity",
             type="log_feature_sparsity",
             metadata=dict(sae_group.cfg.__dict__),
         )
         sparsity_artifact.add_file(log_feature_sparsity_path)
         wandb.log_artifact(sparsity_artifact)
+    return SaveCheckpointOutput(path, log_feature_sparsity_path, log_feature_sparsities)
 
-    return sae_group
-
-
-def wandb_log_suffix(cfg: Any, hyperparams: Any):
-    # Create a mapping from cfg list keys to their corresponding hyperparams attributes
-    key_mapping = {
-        "hook_point_layer": "layer",
-        "l1_coefficient": "coeff",
-        "lp_norm": "l",
-        "lr": "lr",
-    }
 
-    # Generate the suffix by iterating over the keys that have list values in cfg
-    suffix = "".join(
-        f"_{key_mapping.get(key, key)}{getattr(hyperparams, key, '')}"
-        for key, value in vars(cfg).items()
-        if isinstance(value, list)
-    )
-    return suffix
+def _log_feature_sparsity(
+    feature_sparsity: torch.Tensor, eps: float = 1e-10
```

Comment:
A lot of these epsilons are too small to be usable in fp16, Is it OK to increase this to ~1e-7 instead so fp16 works as well?

Reply:
good catch. seems fine to me.